docs/reference/ingest/processors/json.asciidoc

```diff
@@ -10,10 +10,12 @@ Converts a JSON string into a structured JSON object.
 .Json Options
 [options="header"]
 |======
-| Name           | Required  | Default  | Description
-| `field`        | yes       | -        | The field to be parsed.
-| `target_field` | no        | `field`  | The field that the converted structured object will be written into. Any existing content in this field will be overwritten.
-| `add_to_root`  | no        | false    | Flag that forces the serialized json to be injected into the top level of the document. `target_field` must not be set when this option is chosen.
+| Name                  | Required  | Default  | Description
+| `field`               | yes       | -        | The field to be parsed.
+| `target_field`        | no        | `field`  | The field that the converted structured object will be written into. Any existing content in this field will be overwritten.
```

Comment:
It sounds like the "any existing content in this field will be overwritten" statement will no longer be necessarily true.

Reply:
The question is whether or not we want to alter the behavior for target_field. In my initial attempt, I've tried that but it made things quite complex. One reason is that `target_field` allows the source and target to be a scalar whereas `add_to_root` requires the source to be a map/object. Also, today the default behavior differs a lot depending on whether `target_field` or `add_to_root` is set.

The setting `target_field` always overrides the target. When `add_to_root` is set, the fields are kind of merged, but only on the first level. When we introduce a new `merge` option, it would be easier to only let that apply to `add_to_root`, however, that would make it somewhat confusing.

For now, I considered the behavior of the partial merge that applies if `add_to_root` is set to be a bug as it's arguably quite surprising. But I see that this may lead to backward-incompatible behavior if folks rely on the current behavior, as surprising as it may be.

I see three options now:
- Leave it as is and declare the partial merging behavior a bug.
- Introduce `merge_to_root` that only applies for the `add_to_root`, leaving the existing partial merge of `add_to_root` and the overriding behavior of `target_field` intact.
- Introduce `merge` that applies to both `target_field` and `add_to_root`. If set to `false` (default), leaving the existing partial merge of `add_to_root` and the overriding behavior of `target_field` intact.

Reply:
Thanks for the explanation. I'll bring this up at our team meeting and see if we can settle on a preferred option.

Reply:
Based on the outcome of the discussion in the meeting, I've added an `add_to_root_recursive_merge` flag.